### DISCUSSION TITLE
Upgrade/Level-up MEM chips to hold more attackers

### DISCUSSION THREAD

**Discussion by Kutabu** [ochadenas/cpudefense#124] [Ideas]
@cwpute in #119

> Allow MEM chips to level up to hold several attackers: 1 at level 1, 2 at level 2, 3 at level 3 etc… they would only need be emptied when that upgraded limit is reached.
> Obviously, since this is a very powerful upgrade, the price for leveling them up should be very high: 150 bits for the first level-up maybe ? then increasingly pricey.
> If this still is too powerful, you could tie MEM chip max-level to hero upgrades, so that you'd have to first spend cryptocoins on upgrades, then pay a price in bits while playing the levels.
> 
> I had thought about this since a good time now, but MEM chips were already very powerful… but now that they need you to actively empty them, this idea can be considered ! The chips being able to swalow several attackers before needing a reset might even be a double-edged sword, as the player could forget their highly levelled-up chip even needs attention, and let a whole bunch of attackers slip through.
> 
> Such upgrades i thought about while reading about @The-JMFB 's forray into ultra-hard levels in Endless 80 and above. As designed, they are evidently very much fitted for end-game levels: they would serve as even more power-progression and defense potential in later levels, and provide new interesting and anticipated management choices in the form of chip level-ups of an entirely different scale.

---
@The-JMFB in #119
> I'll second that suggestion - a very good option to help balance the swarm effect of later levels!
---

I think your idea of saving several attackers is great! Very useful in the later endless levels.

You would have to decide whether the MEM chips can also be cleared if they are not yet completely full. This would have a significant impact on the gameplay, because a chip that is 4/5 full only catches one attacker in the next wave and then needs time to cool down.

This is very similar to the discussion about the premature releasing of the stored number from ACC chips that we had in #68.

_Side note: After having to empty the MEM chips manually, I had the idea that you could introduce automatic clearing with an upgrade. The higher the upgrade, the faster the chip clears itself. However, the gameplay showed me that this is quite unnecessary because you can always clear them yourself in a short time._

**Comment by ochadenas** [comment]
I would have thought that each tap clears one attacker (and then you must wait for the cool-down). Or, alternatively, a tap clears all of them, but the cool-down time gets multiplied by the count of attackers stored in the chip.

**Comment by Cwpute** [comment]
If each tap should release a single attacker, i believe it'll make players frantically and mindlessly tap on MEM chips, as it'll be the least demanding way of dealing with them.

You second suggestion works better, but its downside is too much: why bother upgrading if the cooldown time is practically the same ?
I could accept an increased cooldown of 50% per MEM level i think, and even then i find quite punitive, but multiplying the cooldown by the number of attackers cleared is unusable as an upgrade… i'd effectively not upgrade my MEM chips if it were the case !


Le 16 mars 2024 14:24:47 GMT+01:00, ochadenas ***@***.***> a écrit :

**Comment by Kutabu** [comment]
1. I like the one-tab-one-attaker idea more and do not share @Cwpute opinion, that it will lead to mindlessly tapping MEM chips. This will not overpower the chip, because you always have to wait the (full) cooldown time and interact with the MEM multiple times.
2. Accumulation the cooldown-time is maybe not so good to the strategic gameplay. This could lead to problems, als the chip is out of order for a long time and the attacking-situation cannot be foreseen for long. On the other hand: Upgrading-MEM is more a lategame thing, therefore the CLK will cut this time anyway.
**More ideas:**
4. If you decide for accumulation of cooldown-time, there could be a reduces summation (inspired by @Cwpute Cwpute). Like:
First attacker: 100% time
Second attacker: +50% time
Third attacker: +25% time (and so on)
6. You cannot clear the MEM chip anymore. Chips cleanes itself after it is "full".

Remarks: There should be an indicator, how full the chip is.

**Comment by ochadenas** [comment]
You're right: Increasing the cool-down doesn't make sense, because the CLK chip will cancel that out anyway. 

I'd like to try with the variant "cool down as before, and attackers must be cleared one-by-one". I think that this is already a huge advantage. Currently, the main weakness of the MEM is that it cannot capture two attackers that are close to each other, even if _on average_ there are not too many of them.

**Comment by Cwpute** [comment]
Capturing an attacker is instantaneous but needs player input for clearing up. Cool-down happens when the chip is manually tapped on by the player.
But... what if MEM chips would only trigger a cooldown when they capture an attacker instead ? During this time, the chip is storing the captured attacker: it can still be tapped on to release previsouly stored attackers, but the one currently being stored will be available for clearing out onyl after the cooldown is completed. Manually clearing up MEM chips would still be needed, but would then be instantaneous. That would allow for MEM chip capacity upgrades without having to fiddle with cooldown time to balance it out!

**Comment by ochadenas** [comment]
Among the numerous ideas and propositions, I have done the following (version 1.33):

- MEM chips can be upgraded up to level 4
- The upgrade cost follow the general principle of "1.5 times the accumulated value". This means that the upgrades cost 18, 45, and 112 bits. Maybe this is too cheap.
- Clearing must be done manually, one slot at a time. There is an optical indication whether you can tap or not.
- The cool down time is unchanged.

Have a look at it once the version gets available and tell me what you think of it ... Perhaps this must be adjusted.

**Comment by Cwpute** [comment]
I happened to have some time over lunch and did a quick test. It's way too powerful 🤡 AND game-changing.

As i expected, i ended up mindlessly tapping the same two MEM chips over and over without a second of strategic thinking, watching attackers waves being diligently swallowed by my allmighty and cuning strategy: placing down exactly two chips and tapping on flashing squares.

I have to come clean because i was the one to request more player interaction, but this is way too much, to the point of becoming an entirely different game altogether, both because you can forgo any strategy in a supposedly slow-paced tower-defense game, and because MEM chips are now so different that they created their own unique little game within the game as a side-effect, on of looking out for and tapping on flashing squares. Your suggestion to have cooldown time multiplied by the number of attackers currently stored now seems much more appealing and calming than the current implementation !

To balance all this, first off: yes, the cost for upgrades is way too cheap. Currently the prices for buying/upgrading are: 12>18>45>172. MEM2 should cost *at least* around what MEM3 costs right now, and MEM3 should then cost what a potential MEM5 would, so around 370bits, again, *at least*.
Another point of regulation i think i hinted at was to block MEM chip levels behind costly hero upgrades, and limit them to MEM3 at most, and even go as far as to only unlock this hero upgrade in Endless.
Why make these so hard to get ? because all those MEM suggestions that were made for very late-game players who already push the boundaries and still want more, to reward them with yet more sense of progression. So even my attempted balance propositions are pretty frail roadblocks in comparison to what they have already achieved and will perhaps be achieving even more easily with this new upgradable chip.

But in the end, this is all impossible to balance because we're trying to acomodate for a game mode (Endless) that's a quick extra feature gifted unto us, built on top of riches (cryptocoins) amassed during the lenght of a previous different quick extra feature (Turbo), and both were wildly unbalanced from the start because that would have been too hard to do right in so little time.
The current Normal mode, the official, hand-crafted mode that catched us by surprise and that we came to love, isn't given nearly as much love. Probably we should focus on that instead of trying so hard a task on such shakey foundations.

Le 20 mars 2024 14:01:24 GMT+01:00, ochadenas ***@***.***> a écrit :

**Comment by Kutabu** [comment]
I have installed 1.33 and am still testing it. Unfortunately I couldn't save the game, so I started in normal mode.

I think the new MEM version is good. Especially the way the "memory status" is displayed. But I don't end up like @Cwpute constantly clicking on MEM chips. They've always been the last defence, even in endless mode. 

As you have already written, the upgrades are very cheap. They should be raised. For the first test, it was okay, but a bit of cost balance might be good. But you are still in the development and better are small changes than overcompensation between the versions.

And I also like @Cwpute's idea of tying the MEM upgrade to a hero that is unlocked after turbo mode, for example. Each hero upgrade = 1 possible upgrade of the MEM chips. This would have less influence on the regular gameplay and would be a real late game upgrade. In other words, exactly when the upgrade function becomes relevant.

**Comment by ochadenas** [comment]
I think both suggestions are good: Make the upgrade price higher for MEM, and require a high-level hero as prerequisite. Will implement this in the next version.

Currently MEM chips are capped at level 4, but that is arbitrary and could be changed easily. Should the max level be higher? It's just that I thought that even more slots would not add to game play.

**Comment by Kutabu** [comment]
I think max upgrade will be limited by price and graphic. On a certain point, you run out of space for memory-blocks at the bottom of the chip.
I suggest: Keep it at 4 and wait for feedback, ob all the endless level 50+ players. They will complain if 4 upgrades are not enough for them ;-)

**Comment by Cwpute** [comment]
I came up with an idea regarding MEM chips, that will kill three birds with one stone, the three birds being:
1. the desire to upgrade MEM chips possibly indefinitely
2. MEM chips currently not taking attacker value into account 
3. some players who want cryptocoins to be affected by more chips

Here it is: MEM chips will have a limited capacity, either in bits or in hex value. They will swallow attackers until this limit is reached, eventually spit out any surplus info that couldn't fit in the form of an attacker, and wait for player input to empty them.
Space already occupied in the chip would be symbolized by a transparent color slowly filling the chip from the bottom up:
![image4](https://github.com/ochadenas/cpudefense/assets/80604512/51d2f5fe-5dd8-4266-b095-fbbd666996fe)
![image3](https://github.com/ochadenas/cpudefense/assets/80604512/1537c8fd-ab8a-418c-8ba0-7771af7b9b9d)
![image4](https://github.com/ochadenas/cpudefense/assets/80604512/1132099c-2982-40f2-8202-e0346e04d8f4)
They would have a medium cooldown between each swallowed attacker, to limit their efficiency without using a clock, and another when it's emptied.
Higher chip levels would upgrade their capacity by 1 or 2 bits, or 1 hexadecimal each. On my mockup i used the MEM chip as we knew it, but its max capacity could be written on it instead, or below its name, so players can appreciate their chip's capacity in relation to uncoming attackers more easily.
Cryptocoins could now be affected by them, since they take attacker value into account.

**Comment by ochadenas** [comment]
Sounds interesting, but quite complex. For the moment at least, I've implemented a simple solution in 1.34: There is a new hero, Leibniz, that becomes available after Grace Hopper and allows for upgrades of MEM chips. And the upgrade prices are now 36 for the 1st upgrade, 90 for the second and 225 for the 3rd. The chip is limited to 4 slots max, but that can easily be changed.
